diff --git a/nbproject/Makefile-Debug.mk b/nbproject/Makefile-Debug.mk
index 5e5027f..0614057 100644
--- a/nbproject/Makefile-Debug.mk
+++ b/nbproject/Makefile-Debug.mk
@@ -52,8 +52,8 @@ OBJECTFILES= \
 CFLAGS=
 
 # CC Compiler Flags
-CCFLAGS=-m32
-CXXFLAGS=-m32
+CCFLAGS=-m$(_BITS)
+CXXFLAGS=-m$(_BITS)
 
 # Fortran Compiler Flags
 FFLAGS=
diff --git a/nbproject/Makefile-Release.mk b/nbproject/Makefile-Release.mk
index ae41bc1..e9cd54e 100644
--- a/nbproject/Makefile-Release.mk
+++ b/nbproject/Makefile-Release.mk
@@ -52,8 +52,8 @@ OBJECTFILES= \
 CFLAGS=
 
 # CC Compiler Flags
-CCFLAGS=-m32
-CXXFLAGS=-m32
+CCFLAGS=-m$(_BITS)
+CXXFLAGS=-m$(_BITS)
 
 # Fortran Compiler Flags
 FFLAGS=
diff --git a/nbproject/Makefile-variables.mk b/nbproject/Makefile-variables.mk
index 3bf56db..af4d99c 100644
--- a/nbproject/Makefile-variables.mk
+++ b/nbproject/Makefile-variables.mk
@@ -3,25 +3,35 @@
 #
 # NOCDDL
 #
+# Fix 32/64 bit compatibility
+ifndef _ARCH
+	_ARCH := $(shell uname -m)
+	export _ARCH
+endif
+ifndef _BITS
+	_BITS := $(shell getconf LONG_BIT)
+	export _BITS
+endif
+
 CND_BASEDIR=`pwd`
 CND_BUILDDIR=build
 CND_DISTDIR=dist
 # Debug configuration
-CND_PLATFORM_Debug=GNU-Linux-x86
-CND_ARTIFACT_DIR_Debug=dist/Debug/GNU-Linux-x86
+CND_PLATFORM_Debug=GNU-Linux-$(_ARCH)
+CND_ARTIFACT_DIR_Debug=dist/Debug/GNU-Linux-$(_ARCH)
 CND_ARTIFACT_NAME_Debug=msed
-CND_ARTIFACT_PATH_Debug=dist/Debug/GNU-Linux-x86/msed
-CND_PACKAGE_DIR_Debug=dist/Debug/GNU-Linux-x86/package
+CND_ARTIFACT_PATH_Debug=dist/Debug/GNU-Linux-$(_ARCH)/msed
+CND_PACKAGE_DIR_Debug=dist/Debug/GNU-Linux-$(_ARCH)/package
 CND_PACKAGE_NAME_Debug=msed.tar
-CND_PACKAGE_PATH_Debug=dist/Debug/GNU-Linux-x86/package/msed.tar
+CND_PACKAGE_PATH_Debug=dist/Debug/GNU-Linux-$(_ARCH)/package/msed.tar
 # Release configuration
-CND_PLATFORM_Release=GNU-Linux-x86
-CND_ARTIFACT_DIR_Release=dist/Release/GNU-Linux-x86
+CND_PLATFORM_Release=GNU-Linux-$(_ARCH)
+CND_ARTIFACT_DIR_Release=dist/Release/GNU-Linux-$(_ARCH)
 CND_ARTIFACT_NAME_Release=msed
-CND_ARTIFACT_PATH_Release=dist/Release/GNU-Linux-x86/msed
-CND_PACKAGE_DIR_Release=dist/Release/GNU-Linux-x86/package
+CND_ARTIFACT_PATH_Release=dist/Release/GNU-Linux-$(_ARCH)/msed
+CND_PACKAGE_DIR_Release=dist/Release/GNU-Linux-$(_ARCH)/package
 CND_PACKAGE_NAME_Release=msed.tar
-CND_PACKAGE_PATH_Release=dist/Release/GNU-Linux-x86/package/msed.tar
+CND_PACKAGE_PATH_Release=dist/Release/GNU-Linux-$(_ARCH)/package/msed.tar
 #
 # include compiler specific variables
 #
diff --git a/nbproject/Makefile-Debug.mk b/nbproject/Makefile-Debug.mk
index 0614057..cadd367 100644
--- a/nbproject/Makefile-Debug.mk
+++ b/nbproject/Makefile-Debug.mk
@@ -21,7 +21,7 @@ FC=gfortran
 AS=as
 
 # Macros
-CND_PLATFORM=GNU-Linux-x86
+CND_PLATFORM=GNU-Linux-$(_ARCH)
 CND_DLIB_EXT=so
 CND_CONF=Debug
 CND_DISTDIR=dist
diff --git a/nbproject/Makefile-Release.mk b/nbproject/Makefile-Release.mk
index e9cd54e..4e09d90 100644
--- a/nbproject/Makefile-Release.mk
+++ b/nbproject/Makefile-Release.mk
@@ -21,7 +21,7 @@ FC=gfortran
 AS=as
 
 # Macros
-CND_PLATFORM=GNU-Linux-x86
+CND_PLATFORM=GNU-Linux-$(_ARCH)
 CND_DLIB_EXT=so
 CND_CONF=Release
 CND_DISTDIR=dist
diff --git a/nbproject/Package-Debug.bash b/nbproject/Package-Debug.bash
index b94d093..fc02abb 100644
--- a/nbproject/Package-Debug.bash
+++ b/nbproject/Package-Debug.bash
@@ -6,7 +6,7 @@
 
 # Macros
 TOP=`pwd`
-CND_PLATFORM=GNU-Linux-x86
+CND_PLATFORM=GNU-Linux-$(_ARCH)
 CND_CONF=Debug
 CND_DISTDIR=dist
 CND_BUILDDIR=build
diff --git a/nbproject/Package-Release.bash b/nbproject/Package-Release.bash
index ae38784..89efdc8 100644
--- a/nbproject/Package-Release.bash
+++ b/nbproject/Package-Release.bash
@@ -6,7 +6,7 @@
 
 # Macros
 TOP=`pwd`
-CND_PLATFORM=GNU-Linux-x86
+CND_PLATFORM=GNU-Linux-$(_ARCH)
 CND_CONF=Release
 CND_DISTDIR=dist
 CND_BUILDDIR=build
