--- pithos/pithos.py	2013-12-05 09:59:17.000000000 -0500
+++ pithos/pithos.py	2014-01-02 02:10:59.456177003 -0500
@@ -185,7 +185,7 @@
         bus.connect("message::eos", self.on_gst_eos)
         bus.connect("message::buffering", self.on_gst_buffering)
         bus.connect("message::error", self.on_gst_error)
-        self.player.connect("notify::volume", self.on_gst_volume)
+        #self.player.connect("notify::volume", self.on_gst_volume)
         self.player.connect("notify::source", self.on_gst_source)
         self.time_format = gst.Format(gst.FORMAT_TIME)
 
